### PR TITLE
fix hang when call flushRequests after resendZkAuth has error

### PR DESCRIPTION
**Pull request by hr3lxphr6j**

### CODE REVIEW

File: zk/conn.go [hunk 1]
```diff
@@ -99,9 +99,10 @@ type Conn struct {
 	closeChan    chan struct{} // channel to tell send loop stop
 
 	// Debug (used by unit tests)
-	reconnectLatch   chan struct{}
-	setWatchLimit    int
-	setWatchCallback func([]*setWatchesRequest)
+	reconnectLatch        chan struct{}
+	setWatchLimit         int
+	setWatchCallback      func([]*setWatchesRequest)
+	resendZkAuthWithError bool
```
Comment (jdef) [inline zk/conn.go]:
```suggestion
	sendDataAfterRequestQueuedTestHook func(*request) error
```

File: zk/conn.go [hunk 2]
```diff
@@ -813,6 +814,13 @@ func (c *Conn) sendData(req *request) error {
 	c.requests[req.xid] = req
 	c.requestsLock.Unlock()
 
+	if req.opcode == opSetAuth && c.resendZkAuthWithError {
```
Comment (jdef) [inline zk/conn.go]:
```suggestion
	var err error
	if c.sendDataAfterRequestQueuedTestHook != nil {
		err = c.sendDataAfterRequestQueuedTestHook(req)
	}
	if err == nil {
		c.conn.SetWriteDeadline(time.Now().Add(c.recvTimeout))
		_, err = c.conn.Write(c.buf[:n+4])
		c.conn.SetWriteDeadline(time.Time{})
	}
```

File: zk/zk_test.go [hunk 3]
```diff
@@ -365,6 +366,45 @@ func TestAuth(t *testing.T) {
...
+	defer ts.Stop()
+	zk, _, err := ts.ConnectAll()
+	if err != nil {
+		t.Fatalf("Connect returned error: %+v", err)
+	}
+	defer zk.Close()
+
+	zk.AddAuth("digest", []byte("user:password"))
+
+	zk.resendZkAuthWithError = true
```
Comment (jdef) [inline zk/zk_test.go]:
```suggestion
	zk.sendDataAfterRequestQueuedTestHook = func(req *request) error {
		if req.opcode == opSetAuth {
			return errors.New("send auth failed")
		}
		return nil
	}
```

File: zk/conn.go [hunk 4]
```diff
@@ -813,6 +814,13 @@ func (c *Conn) sendData(req *request) error {
...
+		err := errors.New("send auth failed")
```
Comment (jdef) [inline zk/conn.go]:
```suggestion
```

File: zk/conn.go [hunk 5]
```diff
@@ -813,6 +814,13 @@ func (c *Conn) sendData(req *request) error {
...
+		req.recvChan <- response{-1, err}
```
Comment (jdef) [inline zk/conn.go]:
```suggestion
```

File: zk/conn.go [hunk 6]
```diff
@@ -813,6 +814,13 @@ func (c *Conn) sendData(req *request) error {
...
+		c.conn.Close()
```
Comment (jdef) [inline zk/conn.go]:
```suggestion
```

File: zk/conn.go [hunk 7]
```diff
@@ -813,6 +814,13 @@ func (c *Conn) sendData(req *request) error {
...
+		return err
```
Comment (jdef) [inline zk/conn.go]:
```suggestion
```

File: zk/conn.go [hunk 8]
```diff
@@ -813,6 +814,13 @@ func (c *Conn) sendData(req *request) error {
...
+	}
```
Comment (jdef) [inline zk/conn.go]:
```suggestion
```

File: zk/conn.go [hunk 9]
```diff
@@ -813,6 +814,13 @@ func (c *Conn) sendData(req *request) error {
...
+
 	c.conn.SetWriteDeadline(time.Now().Add(c.recvTimeout))
```
Comment (jdef) [inline zk/conn.go]:
```suggestion
```

File: zk/conn.go [hunk 10]
```diff
@@ -813,6 +814,13 @@ func (c *Conn) sendData(req *request) error {
...
 	_, err = c.conn.Write(c.buf[:n+4])
```
Comment (jdef) [inline zk/conn.go]:
```suggestion
```

File: zk/conn.go [hunk 11]
```diff
@@ -813,6 +814,13 @@ func (c *Conn) sendData(req *request) error {
...
 	c.conn.SetWriteDeadline(time.Time{})
```
Comment (jdef) [inline zk/conn.go]:
```suggestion
```

**Review by jdef — COMMENTED** [review]
Thanks for the PR. I'd like to see a more generic test hook vs. a special-purpose bool value used here, left some related suggestions.

**Review by jdef — APPROVED** [review]
LGTM.

@mhrabovcin PTAL

**Comment by hr3lxphr6j** [comment]
@mhrabovcin PING

**Review by mhrabovcin — APPROVED** [review]
Thank you! Nice test 🎉

**Pull request merged**